### PR TITLE
DPE-2654 profile-limits option

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

* **main issue**: no fine tune for memory allocation
* configurations are being set without validation
* configurations changes are not triggering changes in the workload

## Solution

* Add `profile-limit-memory` config option
* Usage of data platform `data-models` lib for config validation
* Usage of `rollingOps` lib for workload restart orchestration, when the changed configs requires.
  * Add a method to defer primary to be last one to restart (mitigating primary switchover)

### Chore

* refactored `get_available_memory` to be simpler
* charm typing for `relations/mysql.py`
* typo fix
* usage of config values as properties, provided by `data-models` lib

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -189,22 +200,76 @@ def _on_leader_elected(self, _) -> None:
                 )
         self.unit_peer_data.update({"leader": "true"})
 
+        # Create and set cluster and cluster-set names in the peer relation databag
+        common_hash = generate_random_hash()
+        self.app_peer_data.setdefault(
+            "cluster-name", self.config.cluster_name or f"cluster-{common_hash}"
+        )
+        self.app_peer_data.setdefault("cluster-set-domain-name", f"cluster-set-{common_hash}")
```
Comment (taurus-forever) [inline src/charm.py:208]: Q: will we overwrite `cluster-name` and `cluster-set-domain-name` on the leader move (re-elected)?

**Review by taurus-forever — APPROVED** [review]
Nice. The new code it way more pleasant to read!
One question inside, and I hope the answer there is "no", so LGTM.

File: src/charm.py [hunk 2]
```diff
@@ -189,22 +200,76 @@ def _on_leader_elected(self, _) -> None:
...
-        # Create and set cluster and cluster-set names in the peer relation databag
-        common_hash = generate_random_hash()
-        self.app_peer_data.setdefault(
-            "cluster-name", self.config.get("cluster-name", f"cluster-{common_hash}")
+        if not self.upgrade.idle:
+            # skip when upgrade is in progress
+            # the upgrade already restart the daemon
+            return
+
+        if not self._mysql.is_mysqld_running:
```
Comment (shayancanonical) [inline src/charm.py:225]: this is not a property. should we add `()` to invoke the method?

File: lib/charms/mysql/v0/mysql.py [hunk 3]
```diff
@@ -116,15 +116,17 @@ def wait_until_mysql_connection(self) -> None:
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 49
+LIBPATCH = 50
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:119]: please update to 51 as another PR was merged with libpatch 50

File: src/charm.py [hunk 4]
```diff
@@ -189,22 +200,76 @@ def _on_leader_elected(self, _) -> None:
...
+                    # Wait other units restart first to minimize primary switchover
+                    logger.debug("Primary is waiting for other units to restart")
+                    event.defer()
+                    return
+
+            # persist config to file
+            self._mysql.write_content_to_file(
+                path=MYSQLD_CUSTOM_CONFIG_FILE, content=new_config_content
+            )
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:265]: it's like that to shut the linter - `self.restart.name` is typed ad `AnyStr` and key should be `str`

File: src/config.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        Limited to 63 characters, and must start with a letter and
+        contain only alphanumeric characters, `-`, `_` and `.`
+        """
+        if len(value) > 63:
+            raise ValueError("Cluster name must be less than 63 characters")
+
+        if not value[0].isalpha():
+            raise ValueError("Cluster name must start with a letter")
+
+        if not re.match(r"^[a-zA-Z0-9-_.]*$", value):
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/config.py:87]: not when within the range `[]` - it would otherwise

File: src/charm.py (hunk 2 above)
Comment (paulomach) [reply src/charm.py]: 740b58e

File: lib/charms/mysql/v0/mysql.py (hunk 3 above)
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: 096fe7f2

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -119,56 +118,32 @@ async def get_primary_unit(
...
+    result = await action.wait()
 
-    for unit in ops_test.model.applications[app_name].units:
-        if unit.name == primary_name:
-            return unit
+    primary_unit = None
+    for k, v in result.results["status"]["defaultreplicaset"]["topology"].items():
+        if v["memberrole"] == "primary":
+            unit_name = f"{app_name}/{k.split('-')[-1]}"
+            primary_unit = [unit for unit in units if unit.name == unit_name][0]
```
[the remark this answers is not in the document]
Comment (paulomach) [reply tests/integration/helpers.py:141]: [it does](https://ops.readthedocs.io/en/latest/index.html#ops.Model.get_unit), but not ops_test

**Review by shayancanonical — APPROVED** [review]

File: lib/charms/mysql/v0/mysql.py [hunk 7]
```diff
@@ -1141,6 +1149,11 @@ def set_dynamic_variable(
         """
         if not instance_address:
             instance_address = self.instance_address
+
+        # escape variable values when needed
+        if not re.match(r"^[0-9,a-z,A-Z$_]+$", value):
+            value = f"`{value}`"
+
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:1156]: actually, we cannot escape always - number will fail. rolledback on ea0ecb9

**Comment by paulomach** [comment]
merging besides the CI error - test passed previously, failing due disk full error. Deal with it on another PR

**Pull request merged**